src/Draco.FormatterEngine/Scope.cs

```diff
@@ -0,0 +1,184 @@
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Draco.Compiler.Internal.Syntax.Formatting;
+
+/// <summary>
+/// Represent an indentation level in the code.
+/// </summary>
+public sealed class Scope
+{
+    private readonly string? indentation;
+    private readonly (IReadOnlyList<TokenMetadata> tokens, int indexOfLevelingToken)? levelingToken;
+    private readonly FormatterSettings settings;
+
+    [MemberNotNullWhen(true, nameof(levelingToken))]
+    [MemberNotNullWhen(false, nameof(indentation))]
+    private bool DrivenByLevelingToken => this.levelingToken.HasValue;
+
+    private Scope(Scope? parent, FormatterSettings settings, FoldPriority foldPriority)
+    {
+        this.Parent = parent;
+        this.settings = settings;
+        this.FoldPriority = foldPriority;
+    }
+
+    /// <summary>
+    /// Create a scope that add a new indentation level.
+    /// </summary>
+    /// <param name="parent">The parent scope.</param>
+    /// <param name="settings">The settings of the formatter.</param>
+    /// <param name="foldPriority">The fold priority of the scope.</param>
+    /// <param name="indentation">The indentation this scope will add.</param>
+    public Scope(Scope? parent, FormatterSettings settings, FoldPriority foldPriority, string indentation) 
+        : this(parent, settings, foldPriority)
+    {
+        this.indentation = indentation;
+    }
+
+    /// <summary>
+    /// Create a scope that override the indentation level and follow the position of a token instead.
+    /// </summary>
+    /// <param name="parent">The parent scope.</param>
+    /// <param name="settings">The settings of the formatter.</param>
+    /// <param name="foldPriority">The fold priority of the scope.</param>
+    /// <param name="levelingToken">The list of the tokens of the formatter and the index of the token to follow the position.</param>
+    public Scope(Scope? parent, FormatterSettings settings, FoldPriority foldPriority, (IReadOnlyList<TokenMetadata> tokens, int indexOfLevelingToken) levelingToken)
+        : this(parent, settings, foldPriority)
+    {
+        this.levelingToken = levelingToken;
+    }
+
+    /// <summary>
+    /// The parent scope of this scope.
+    /// </summary>
+    public Scope? Parent { get; }
+
+    /// <summary>
+    /// Represent if the scope is materialized or not.
+    /// An unmaterialized scope is a potential scope, which is not folded yet.
+    /// <code>items.Select(x => x).ToList()</code> have an unmaterialized scope.
+    /// It can be materialized like:
+    /// <code>
+    /// items
+    ///     .Select(x => x)
+    ///     .ToList()
+    /// </code>
+    /// </summary>
+    public Future<bool> IsMaterialized { get; } = new Future<bool>();
```

Comment:
Concept isn't 100% clear, does unmaterialized mean not folded?

Reply:
it's if the scope exists or not.  
It could be renamed Exists ?